Overhaul notice conditions

Cyperghost opened the pull request:
TODO

- [x] Migration of old data
- [x] Info that rebuild data for notice is required
  - [x] In the ACP Dashboard box
  - [x] In the list of notices

Cyperghost committed:
Migrate notices to new condition system

Cyperghost committed:
Add DayOfWeek and NotDayOfWeek condition types

Cyperghost committed:
Add CSS class name selection field with custom option support

Cyperghost committed:
Add functionality to reset dismissed notices in form

Cyperghost committed:
Add migration functionality for notice

Cyperghost committed:
Implement migration support for request condition types

Cyperghost committed:
Migrate `NoticeCacheBuilder` to an eager cache

dtdesign commented on wcfsetup/install/files/lib/data/notice/NoticeAction.class.php:
```diff
@@ -134,4 +147,22 @@ public function update()
+    #[\Override]
+    public function getPackageID(): int
+    {
+        return PACKAGE_ID;
```
Do not do this! `PACKAGE_ID` maps to the currently loaded app which means that calling this action from anywhere else but the notice form in the admin panel has the risk of binding the phrases to the active app. Uninstalling that app would then remove the phrases too.

Notices are tied to the core, thus you should set this to the explicit value of `1`.

dtdesign commented on wcfsetup/install/files/lib/system/condition/type/IGlobalConditionType.class.php:
```diff
@@ -0,0 +1,20 @@
+ * @template TFilter
+ * @extends  IConditionType<TFilter>
+ */
+interface IGlobalConditionType extends IConditionType
```
I think `IContextualConditionType` is a more fitting name for this interface.

dtdesign commented on wcfsetup/install/files/lib/system/condition/type/request/NotDayOfWeekRequestConditionType.class.php:
```diff
@@ -0,0 +1,94 @@
+        // We must remove all selected week of days to convert the previous condition from an “or” to an “and” condition.
+        $daysOfWeeks = \array_diff_key(DateUtil::getWeekDays(), $daysOfWeeks);
+
+        foreach ($daysOfWeeks as $dayOfWeek => $_) {
```
```suggestion
        foreach (\array_keys($daysOfWeeks) as $dayOfWeek) {
```

Cyperghost committed:
Update wcfsetup/install/files/lib/system/condition/type/request/NotDayOfWeekRequestConditionType.class.php

Co-authored-by: Alexander Ebert <ebert@woltlab.com>

Cyperghost committed:
Use `1` instead of `PACKAGE_ID`, to ensure that the correct package ID is set

Cyperghost committed:
Rename `IGlobalConditionType` to `IContextualConditionType`

Cyperghost committed:
Remove unnecessary `CombinedConditionProvider`.
Implement the function `withConditionsFrom()` to transfer conditions from another provider.

Cyperghost committed:
The active page must be a multiple selection so that an “or” condition is possible.

Cyperghost committed:
Add the information to the ACP Status Message box that the Rebuild Data must be performed for notes.

Cyperghost committed:
Grouped condition types (#6387)

* Group the condition types and make it possible to open/close the groups in the input field.

* Minor code fixes

* Remove the collapse feature, improve visuals of categories

* Simplify the filtering, use native `hidden` property instead

---------

Co-authored-by: Alexander Ebert <ebert@woltlab.com>
